crypto/fipsmodule/cpucap/cpu_aarch64.c

```diff
@@ -82,11 +82,20 @@ uint64_t armv8_enable_dit(void) {
 }
 
 void armv8_restore_dit(volatile uint64_t *original_dit) {
-  if (CRYPTO_is_ARMv8_DIT_capable() && *original_dit != 1) {
+  if (*original_dit != 1 && CRYPTO_is_ARMv8_DIT_capable()) {
     // Encoding of "msr dit, #0"
     __asm__ volatile(".long 0xd503405f");
   }
 }
-#endif  // MAKE_DIT_AVAILABLE && !OPENSSL_WINDOWS
+
+void armv8_disable_dit(void) {
+  OPENSSL_armcap_P &= ~ARMV8_DIT_ALLOWED;
```

Comment:
here and for `armv8_enable_dit`, is this operation thread-safe? not sure what the `HIDDEN` macro expands to, but it's [not clear](https://github.com/aws/aws-lc/blob/c28d0cb27759b09eb305a98f822c0223eeb2c727/crypto/fipsmodule/cpucap/cpucap.c#L84) that `OPENSSL_armcap_P` is thread-local.

Reply:
`OPENSSL_armcap_P` is a global variable of type `uint32_t` which is shared a cross the entire address space. 

Firstly, since the memory is shared it doesn't allow a thread to control whether it wants to enable DIT mode or not; its choice can be overwritten by any other thread.

Secondly, multiple writes/reads can now occur concurrently on `OPENSSL_armcap_P` since it's mostly just referenced directly and there therefore aren't any synchonisation. This was fine because it was only really readable after the initial construction of the memory content (the initialisation of the arm capability vector). This PR breaks that assumption.

Finally, is this a "threading issue". Well, yes and no. One thread can read the value while another writes to it. The writing operation only mutate a single byte of the 32-bit value. So, the value when read is either the bit set or not. It won't be a bogus value. So, no. On the other hand, from the first point I made above, any thread can mutate the value. So one thread can enable, while another immediately afterwards can disable the DIT. But that might just be what we can build - DIT is a process global CPU control no?

I think we weakly considered `OPENSSL_armcap_P` nominally `const` after the initial initialisation. It might not be the ideal memory to store this information.

Reply:
>  from the first point I made above, any thread can mutate the value. So one thread can enable, while another immediately afterwards can disable the DIT.

yes, this is my concern.

> But that might just be what we can build - DIT is a process global CPU control no?

My understanding from nevine's description is that it can be configured per-thread. I suppose we could leave it up to the caller to not call these functions from >1 thread, and document this limitation very clearly. I don't think this would be an issue for ACCP, as we'd just set or unset DIT once on initialization, but it could be a concern for other consumers.